ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java

```diff
@@ -29,6 +29,47 @@ public class MySqlNamedBlobDbConfig {
   public static final String  QUERY_STALE_DATA_MAX_RESULTS = PREFIX + "query.stale.data.max.results";
   public static final String  STALE_DATA_RETENTION_DAYS = PREFIX + "stale.data.retention.days";
   public static final String TRANSACTION_ISOLATION_LEVEL = PREFIX + "transaction.isolation.level";
+  public static final String LIST_NAMED_BLOBS_SQL = "list.named.blobs.sql";
+
+  /**
+   * List named-blobs query.
+   * The inner query selects blobs based on the filters, and groups them on blob_name.
+   * In each group, it selects the blob with the highest version.
+   * The outer query filters out blobs that have been deleted, and return the latest blob in each group that is ready.
+   */
+  @Config(LIST_NAMED_BLOBS_SQL)
+  public static final String DEFAULT_LIST_NAMED_BLOBS_SQL = ""
+      + " WITH "
+      + "  BlobsAllVersion AS ( "
+      + "   SELECT blob_name, blob_id, version, deleted_ts, blob_size, modified_ts "
+      + "   FROM named_blobs_v2 "
+      + "   WHERE account_id = ? " // 1
+      + "     AND container_id = ? " // 2
+      + "     AND blob_state = %1$s "
+      + "     AND blob_name LIKE ? " // 3
+      + "     AND blob_name >= ? " // 4
+      + "     AND (deleted_ts IS NULL OR deleted_ts > %2$s) "
+      + " ), "
+      + " BlobsMaxVersion AS ( "
+      + "   SELECT blob_name, MAX(version) as version "
+      + "   FROM named_blobs_v2 "
+      + "   WHERE account_id = ? " // 5
+      + "     AND container_id = ? " // 6
+      + "     AND blob_state = %1$s "
+      + "     AND blob_name LIKE ? " // 7
+      + "     AND blob_name >= ? " // 8
+      + "   GROUP BY blob_name "
```

Comment:
I think GROUP BY logic operates independently of the WHERE clause, this GROUP BY account_id, container_id, blob_name ensures that the aggregation (e.g., MAX(version)) respects the boundaries of each account and container. Again, we are not the expert of mysql queries, and you might want to add some test case like adding named blob for different account and container and test the list. Even though you use where to select the corresponding account and container, it might not give you the right max version based on my understanding.

Reply:
I saw you only select the blob_name and  MAX(version), does not select the accountId, so sql won't be confused.

Reply:
Group-by executes after filters in the query plan because that is the least cost way of executing the query.